README.md
# yw2xtg - XPress tagged text export from yWriter projects

For more information, see the [project homepage](https://peter88213.github.io/yw2xtg) with description and download instructions.

## Development

*yw2xtg* is organized as an Eclipse PyDev project. The official release branch on GitHub is *main*.

*yw2xtg* depends on the [pywriter](https://github.com/peter88213/PyWriter) library which must be present in your file system. It is organized as an Eclipse PyDev project. The official release branch on GitHub is *main*.

### Mandatory directory structure for building the application script

```
.
├── PyWriter/
│   └── src/
│       └── pywriter/
└── yw2xtg/
    ├── src/
    ├── test/
    └── tools/ 
        └── build.xml
```

### Conventions

See https://github.com/peter88213/PyWriter/blob/main/docs/conventions.md

## Development tools

- [Python](https://python.org) version 3.10.
- [Eclipse IDE](https://eclipse.org) with [PyDev](https://pydev.org) and *EGit*.
- *Apache Ant* is used for building the application.

## Credits

- The icons are made using the free *Pusab* font by Ryoichi Tsunekawa, [Flat-it](http://flat-it.com/).
- User *Hunter_71* presented the number to English conversion algorithm on [stack overflow](https://stackoverflow.com/a/51849443).
- User *Aristide* presented the integer to roman numeral conversion on [stack overflow](https://stackoverflow.com/a/47713392).


## License

yw2xtg is distributed under the [MIT
License](http://www.opensource.org/licenses/mit-license.php).


docs/changelog.md
[Project home page](index) > Changelog

------------------------------------------------------------------------

## Changelog

### Planned features

See the [GitHub "Features" project](https://github.com/users/peter88213/projects/4/views/1).

### v1.3.0

- Rename the XPress tag that replaces "bold": "@Small caps" --> "@Strong emphasis"

Based on PyWriter v12.14.0

### v1.2.3

- Library update.

Based on PyWriter v12.4.1

### v1.2.2

- Reduce the memory use by discarding the docstrings on building.

Based on PyWriter v12.1.2

### v1.2.1

- Remove "shebang" lines to make the scripts run with Python 3.11 under Windows.

Based on PyWriter v9.0.5

### v1.2.0

- Code optimization and library update. 

Based on PyWriter v9.0.4

### v1.1.2

- Library upgrade.
- Convert language codes.
- Error handling via exceptions instead of returned messages.

Based on PyWriter v8.0.8

### v1.0.10

- Update the PyWriter library for future Python versions.

Based on PyWriter v7.14.2

### v1.0.9

- Update the icon set.
- Refactor the code.

Based on PyWriter v7.4.1

### v1.0.8

- Set own window icon.

Based on PyWriter v7.4.1

### v1.0.7 Optional release

- Code refactoring and library update.

Based on PyWriter v7.2.1

### v1.0.6 Update setup script

- Change the working dir to the script dir on startup in order to avoid "file not found" error.

Based on PyWriter v5.18.0

### v1.0.5 Improved setup

- Catch exceptions in the setup script.

Based on PyWriter v5.18.0

### v1.0.4 Bugfix release

- Fix and refactor inline code removal.

Based on PyWriter v5.16.1

### v1.0.3 Consider inline code

- Remove inline code when exporting.

Based on PyWriter v5.12.5

### v1.0.2

- Improve code and documentation quality.

Based on PyWriter v5.6.1

### v1.0.1

- Improve code and documentation quality.

Based on PyWriter v5.0.2

### v1.0.0

- Fix a regression from v0.12.2 where paragraph indenting doesn't work when exporting to chapters.
- Fix a bug where "To do" chapters cause an exception.
- Process non-ASCII characters in the configuration file. 
- Change the default character style closing tag.
- Add $AuthorBio placeholder.
- Improve the overall code quality.

Based on PyWriter v5.0.0

### v0.12.2 Optional update

Refactor the code.

Based on PyWriter v3.32.2

### v0.12.1 Optional update

Refactor the code.

Based on PyWriter v3.32.2

### v0.12.0 Indent paragraphs that begin with "> "

Based on PyWriter v3.32.2

### v0.10.0 Do not indent paragraphs after blank lines

Please note: The entries in the yw2xtg configuration file have changed.
If you use a custom configuration, do update it manually.

Based on PyWriter v3.32.2

### v0.8.7 Support non-Windows OS

- Move installation and configuration to another location (see instructions for use).

Based on PyWriter v3.28.1

### v0.8.6 Enable non-Windows operation 

- Catch an exception that is thrown when evaluating a Windows environment variable under a non-Windows OS.

Based on PyWriter v3.28.1

### v0.8.5 Optional release

Define the priority of the configurations if there are several.

Configuration and template sources:

1. `yw2xtg` subdirectory in the yWriter project directory.
2. `~\AppData\Roaming\PyWriter\yw2xtg\config`
3. Default values.

If a template file or a configuration entry is missing, *yw2xtg* uses the lower priority source as a fallback. 

Based on PyWriter v3.18.0

### v0.8.4 Bugfix release

This release is strongly recommended.
Fix a regression from PyWriter v3.12.5. causing a crash if a scene has an 
hour, but no minute set.

Based on PyWriter v3.16.4

### v0.8.3 No automatic shortcut creation

- Due to sporadic security warnings, the automatic shortcut creation during installation is removed. The user is now guided to create the application shortcut manually. 
- Refactor: Use standard function for reading boolean values from the INI file.

Based on PyWriter v3.16.0

### v0.8.2 Use backup configuration

If there is no configuration data in the project file, 
data stored in the installation folder is used prior to 
the script's default configuration data.

Based on PyWriter v3.12.7

### v0.8.1 Include installation script

**install.bat** installs the script for the local user and creates a 
shortcut on the desktop.

Based on PyWriter v3.12.7

### v0.8.0 Extend configuration

Move the "ope file per chapter" option from the command line arguments 
to the configuration file.
The chapter files are written to the XTG_chapters subdirectory.

**Note:**

When updating from v0.6, the per_chapter line must be added to existing 
configuration files.

Based on PyWriter v3.12.7

### v0.6.0 Create chapter files

Optionally, one file per chapter is created.

Based on PyWriter v3.12.7

### v0.4.1 Implement a silent mode

Based on PyWriter v3.12.6.

### v0.4.0 Extend the configuration file

- Add OPTIONS section to the configuration.
- Make some text processing steps optional.
- Escape XPress Tags code-specific characters.

Based on PyWriter v3.12.6.

### v0.2.1 Modify default values

Based on PyWriter v3.12.6.

### v0.2.0 Beta test release 

Based on PyWriter v3.12.6.



src/pywxtg/xtg_file.py
"""Provide a class for XPress tagged file processing. 

Copyright (c) 2022 Peter Triesberger
For further information see https://github.com/peter88213/yw2xtg
Published under the MIT License (https://opensource.org/licenses/mit-license.php)
"""
import shutil
import os
import re
from string import Template

from pywriter.pywriter_globals import ERROR
from pywriter.file.file_export import FileExport


class XtgFile(FileExport):
    """XPress tagged file representation.
    """

    DESCRIPTION = 'XPress tagged file'
    EXTENSION = '.XTG'
    SUFFIX = ''
    _XTG_OUT = 'XTG_Chapters'

    def __init__(self, filePath, **kwargs):
        super().__init__(filePath)

        self._fileHeader = kwargs['file_header']
        self._partTemplate = kwargs['part_template']
        self._chapterTemplate = kwargs['chapter_template']
        self._firstSceneTemplate = kwargs['first_scene_template']
        self._sceneTemplate = kwargs['scene_template']
        self._appendedSceneTemplate = kwargs['appended_scene_template']
        self._sceneDivider = kwargs['scene_divider']

        self._tagFirstParagraph = kwargs['first_paragraph']
        self._tagIndentedParagraph = kwargs['indented_paragraph']
        self._tagOtherParagraph = kwargs['other_paragraph']
        self._tagItalic = kwargs['italic']
        self._tagItalic0 = kwargs['italic0']
        self._tagBold = kwargs['bold']
        self._tagBold0 = kwargs['bold0']
        self._tagAcronym = kwargs['acronym']
        self._tagAcronym0 = kwargs['acronym0']
        self._tagFigure = kwargs['figure']
        self._tagFigure0 = kwargs['figure0']

        self._adjustDigits = kwargs['adjust_digits']
        self._spacePoints = kwargs['space_points']
        self._perChapter = kwargs['per_chapter']

    def _convert_from_yw(self, text, quick=False):
        """Convert yw7 markup to Markdown.
        """
        XTG_REPLACEMENTS = [
            # Escape XPress Tags code-specific characters.
            ('@', '┌┐@>'),
            ('<', '┌┐<>'),
            ('\\', '┌┐\\>'),
            ('┌┐', '<\\'),
        ]
        
        if quick:
            # Just clean up a one-liner without sophisticated formatting.
            
            try:
    
                for yw, xt in XTG_REPLACEMENTS:
                    text = text.replace(yw, xt)
                    
                return text
    
            except AttributeError:
                return ''
                        
        XTG_REPLACEMENTS.extend([
            # Replace yWriter tags with XPress tags.
            ('[i]', self._tagItalic),
            ('[/i]', self._tagItalic0),
            ('[b]', self._tagBold),
            ('[/b]', self._tagBold0),
            ('  ', ' '),
            # Format paragraphs.
            ('\n\n', f'\r\r{self._tagFirstParagraph}'),
            ('\n', f'\n{self._tagOtherParagraph}'),
            ('\r', '\n'),
        ])

        try:

            for yw, xt in XTG_REPLACEMENTS:
                text = text.replace(yw, xt)

        except AttributeError:
            return ''

        #--- Encode non-breaking spaces.

        text = text.replace('\xa0', '<\\!p>')

        #--- Remove highlighting, alignment,
        # strikethrough, and underline tags.

        text = re.sub('\[\/*[h|c|r|s|u]\d*\]', '', text)

        #--- Remove comments.

        text = re.sub('\/\*.+?\*\/', '', text)

        #--- Adjust digit-separating blanks.

        if self._adjustDigits:
            text = re.sub('(\d) (\d)', '\\1<\\![>\\2', text)

        #--- Space digit-separating points.

        if self._spacePoints:
            text = re.sub('(\d+)\.', '\\1.<\\![>', text)
            text = text.replace('<\\![> ', ' ')

        #--- Assign "figure" style.

        text = re.sub('(\d+)', f'{self._tagFigure}\\1{self._tagFigure0}', text)

        #--- Assign "acronym" style.

        text = re.sub('([A-ZÄ-Ü]{2,})', f'{self._tagAcronym}\\1{self._tagAcronym0}', text)

        return text

    def _get_chapterMapping(self, chId, chapterNumber):
        """Return a mapping dictionary for a chapter section. 
        """

        ROMAN = [
            (1000, "M"),
            (900, "CM"),
            (500, "D"),
            (400, "CD"),
            (100, "C"),
            (90, "XC"),
            (50, "L"),
            (40, "XL"),
            (10, "X"),
            (9, "IX"),
            (5, "V"),
            (4, "IV"),
            (1, "I"),
        ]

        def number_to_roman(n):
            """Return n as a Roman number.
            Credit goes to the user 'Aristide' on stack overflow.
            https://stackoverflow.com/a/47713392
            """

            result = []

            for (arabic, roman) in ROMAN:
                (factor, n) = divmod(n, arabic)
                result.append(roman * factor)

                if n == 0:
                    break

            return "".join(result)

        TENS = {30: 'thirty', 40: 'forty', 50: 'fifty',
                60: 'sixty', 70: 'seventy', 80: 'eighty', 90: 'ninety'}
        ZERO_TO_TWENTY = (
            'zero', 'one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine', 'ten',
            'eleven', 'twelve', 'thirteen', 'fourteen', 'fifteen', 'sixteen', 'seventeen', 'eighteen', 'nineteen', 'twenty'
        )

        def number_to_english(n):
            """Return n as a number written out in English.
            Credit goes to the user 'Hunter_71' on stack overflow.
            https://stackoverflow.com/a/51849443
            """

            if any(not x.isdigit() for x in str(n)):
                return ''

            if n <= 20:
                return ZERO_TO_TWENTY[n]

            elif n < 100 and n % 10 == 0:
                return TENS[n]

            elif n < 100:
                return f'{number_to_english(n - (n % 10))} {number_to_english(n % 10)}'

            elif n < 1000 and n % 100 == 0:
                return f'{number_to_english(n / 100)} hundred'

            elif n < 1000:
                return f'{number_to_english(n / 100)} hundred {number_to_english(n % 100)}'

            elif n < 1000000:
                return f'{number_to_english(n / 1000)} thousand {number_to_english(n % 1000)}'

            return ''

        chapterMapping = super()._get_chapterMapping(chId, chapterNumber)

        if chapterNumber:
            chapterMapping['ChNumberEnglish'] = number_to_english(chapterNumber).capitalize()
            chapterMapping['ChNumberRoman'] = number_to_roman(chapterNumber)

        else:
            chapterMapping['ChNumberEnglish'] = ''
            chapterMapping['ChNumberRoman'] = ''

        if self.chapters[chId].suppressChapterTitle:
            chapterMapping['Title'] = ''

        return chapterMapping

    def _get_chapters(self):
        """Process the chapters and nested scenes.
        Return a list of strings, or a message, depending on the _perChapter variable.
        Extend the superclass method for the 'document per chapter' option.
        """

        if not self._perChapter:
            return super()._get_chapters()

        # Create a directory for the output. Delete contents, if exist.

        xtgDir = os.path.dirname(self.filePath)

        if xtgDir == '':
            xtgDir = '.'

        xtgDir = f'{xtgDir}/{self._XTG_OUT}'

        if os.path.isdir(xtgDir):
            shutil.rmtree(xtgDir)

        os.makedirs(xtgDir)

        chapterNumber = 0
        sceneNumber = 0
        wordsTotal = 0
        lettersTotal = 0

        for chId in self.srtChapters:
            lines = []

            dispNumber = 0

            if not self._chapterFilter.accept(self, chId):
                continue

            # The order counts; be aware that "Todo" and "Notes" chapters are
            # always unused.

            # Has the chapter only scenes not to be exported?

            sceneCount = 0
            notExportCount = 0
            doNotExport = False
            template = None

            for scId in self.chapters[chId].srtScenes:
                sceneCount += 1

                if self.scenes[scId].doNotExport:
                    notExportCount += 1

            if sceneCount > 0 and notExportCount == sceneCount:
                doNotExport = True

            if self.chapters[chId].chType == 2:
                # Chapter is "ToDo" type (implies "unused").

                if self._todoChapterTemplate != '':
                    template = Template(self._todoChapterTemplate)

            elif self.chapters[chId].chType == 1:
                # Chapter is "Notes" type (implies "unused").

                if self._notesChapterTemplate != '':
                    template = Template(self._notesChapterTemplate)

            elif self.chapters[chId].isUnused:
                # Chapter is "really" unused.

                if self._unusedChapterTemplate != '':
                    template = Template(self._unusedChapterTemplate)

            elif self.chapters[chId].oldType == 1:
                # Chapter is "Info" type (old file format).

                if self._notesChapterTemplate != '':
                    template = Template(self._notesChapterTemplate)

            elif doNotExport:

                if self._notExportedChapterTemplate != '':
                    template = Template(self._notExportedChapterTemplate)

            elif self.chapters[chId].chLevel == 1 and self._partTemplate != '':
                template = Template(self._partTemplate)

            else:
                template = Template(self._chapterTemplate)
                chapterNumber += 1
                dispNumber = chapterNumber

            if template is not None:
                lines.append(template.safe_substitute(self._get_chapterMapping(chId, dispNumber)))

            # Process scenes.

            sceneLines, sceneNumber, wordsTotal, lettersTotal = self._get_scenes(
                chId, sceneNumber, wordsTotal, lettersTotal, doNotExport)
            lines.extend(sceneLines)

            # Process chapter ending.

            template = None

            if self.chapters[chId].chType == 2:

                if self._todoChapterEndTemplate != '':
                    template = Template(self._todoChapterEndTemplate)

            elif self.chapters[chId].chType == 1:
                
                if self._notesChapterEndTemplate != '':
                    template = Template(self._notesChapterEndTemplate)

            elif self.chapters[chId].isUnused:

                if self._unusedChapterEndTemplate != '':
                    template = Template(self._unusedChapterEndTemplate)

            elif self.chapters[chId].oldType == 1:

                if self._notesChapterEndTemplate != '':
                    template = Template(self._notesChapterEndTemplate)

            elif doNotExport:

                if self._notExportedChapterEndTemplate != '':
                    template = Template(self._notExportedChapterEndTemplate)

            elif self._chapterEndTemplate != '':
                template = Template(self._chapterEndTemplate)

            if template is not None:
                lines.append(template.safe_substitute(self._get_chapterMapping(chId, dispNumber)))

            if not lines:
                continue

            text = f'{self._fileHeader}{"".join(lines)}'

            # Fix the tags of indented paragraphs.
            # This is done here to include the scene openings.

            text = re.sub('\n\@.+?:\> ', f'\n{self._tagIndentedParagraph}', text)
            xtgPath = f'{xtgDir}/{dispNumber:04}_{self.chapters[chId].title}{self.EXTENSION}'

            try:
                with open(xtgPath, 'w', encoding='utf-8') as f:
                    f.write(text)

            except:
                return(f'{ERROR}Cannot write "{os.path.normpath(xtgPath)}".')

        return 'All chapters written.'

    def _get_text(self):
        """Assemble the whole text applying the templates.
        Return a string to be written to the output file.
        Override the superclass.
        """
        lines = self._get_fileHeader()
        lines.extend(self._get_chapters())
        text = ''.join(lines)

        # Fix the tags of indented paragraphs.
        # This is done here to include the scene openings.

        text = re.sub('\n\@.+?:\> ', f'\n{self._tagIndentedParagraph}', text)
        return text

    def write(self):
        """Create a template-based output file. 
        Return a message beginning with the ERROR constant in case of error.
        Extend the superclass method for the 'document per chapter'
        option.
        """
        if self._perChapter:
            return self._get_chapters()

        else:
            return super().write()
